vsphere/virtual_machine_config_structure.go

```diff
@@ -0,0 +1,664 @@
+package vsphere
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"reflect"
+
+	"github.com/hashicorp/terraform/helper/logging"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/terraform-providers/terraform-provider-vsphere/vsphere/internal/helper/structure"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+var virtualMachineResourceAllocationTypeValues = []string{"cpu", "memory"}
+
+var virtualMachineVirtualExecUsageAllowedValues = []string{
+	string(types.VirtualMachineFlagInfoVirtualExecUsageHvAuto),
+	string(types.VirtualMachineFlagInfoVirtualExecUsageHvOn),
+	string(types.VirtualMachineFlagInfoVirtualExecUsageHvOff),
+}
+
+var virtualMachineVirtualMmuUsageAllowedValues = []string{
+	string(types.VirtualMachineFlagInfoVirtualMmuUsageAutomatic),
+	string(types.VirtualMachineFlagInfoVirtualMmuUsageOn),
+	string(types.VirtualMachineFlagInfoVirtualMmuUsageOff),
+}
+
+var virtualMachineSwapPlacementAllowedValues = []string{
+	string(types.VirtualMachineConfigInfoSwapPlacementTypeInherit),
+	string(types.VirtualMachineConfigInfoSwapPlacementTypeVmDirectory),
+	string(types.VirtualMachineConfigInfoSwapPlacementTypeHostLocal),
+}
+
+var virtualMachineFirmwareAllowedValues = []string{
+	string(types.GuestOsDescriptorFirmwareTypeBios),
+	string(types.GuestOsDescriptorFirmwareTypeEfi),
+}
+
+// getWithRestart fetches the resoruce data specified at key. If the value has
+// changed, a reboot is flagged in the virtual machine by setting
+// reboot_required to true.
+func getWithRestart(d *schema.ResourceData, key string) interface{} {
+	if d.HasChange(key) {
+		log.Printf("[DEBUG] %s: Resource argument %q requires a VM restart", resourceVSphereVirtualMachineIDString(d), key)
+		d.Set("reboot_required", true)
+	}
+	return d.Get(key)
+}
+
+// getBoolWithRestart fetches a *bool for the resource data item specified at
+// key. If the value has changed, a reboot is flagged in the virtual machine by
+// setting reboot_required to true.
+//
+// This function always returns at least false, even if a value is unspecified.
+func getBoolWithRestart(d *schema.ResourceData, key string) *bool {
+	if d.HasChange(key) {
+		d.Set("reboot_required", true)
+	}
+	return structure.GetBool(d, key)
+}
+
+// schemaVirtualMachineConfigSpec returns schema items for resources that
+// need to work with a VirtualMachineConfigSpec.
+func schemaVirtualMachineConfigSpec() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		// VirtualMachineBootOptions
+		"boot_delay": {
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Description: "The number of milliseconds to wait before starting the boot sequence.",
+		},
+		"efi_secure_boot_enabled": {
```

Comment:
🤔 not a blocker, more a design thought for the future: you could probably combine these into a sub-object and infer that EFI's required based on the block existing:

```
efi {
  secure_boot = true
}
```

Reply:
Yeah - although I grouped certain workflows in a nested structure this time (ie: `customize`), I'd almost say that this one didn't make sense, as the options were in two completely separate places - secure boot is in [`VirtualMachineBootOptions`](https://pubs.vmware.com/vsphere-6-5/topic/com.vmware.wssdk.apiref.doc/vim.vm.BootOptions.html), while EFI itself is the `firmware` property in [`VirtualMachineConfigSpec`](https://pubs.vmware.com/vsphere-6-5/topic/com.vmware.wssdk.apiref.doc/vim.vm.ConfigSpec.html#field_detail). So while mapping this out, it would probably look a bit awkward to throw this in a sub-resource, especially when most other options aren't. To me, this would make more sense if EFI and EFI-specific options were all groped up into their own data object in the API.